src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java

```diff
@@ -32,56 +40,75 @@ public static void main(String[] args) {
             .addAction(new UpcaseFieldsAction().addField("Country"));
         
         pipeline.run();
+        
+        System.out.println("Generated Code:" + System.lineSeparator());
+        System.out.println(pipeline.getJavaCode().getSource());
     }
     
-}
+    static class UpcaseFieldsAction extends PipelineAction {
 
-class UpcaseFieldsAction extends PipelineAction {
+        private final List<String> fieldNameList;
 
-    private final List<String> fieldNameList;
+        protected UpcaseFieldsAction() {
+            super("transform-upcase-fields", "Upper Case Fields");
+            fieldNameList = new ArrayList<>();
+        }
 
-    protected UpcaseFieldsAction() {
-        super("transform-upcase-fields", "Upper Case Fields");
-        fieldNameList = new ArrayList<>();
-    }
+        @Override
+        public DataReader apply(DataReader reader) throws Throwable {
+            TransformingReader transformingReader = new TransformingReader(reader)
+                    .add(new BasicFieldTransformer(fieldNameList.toArray(new String[fieldNameList.size()])).upperCase());
+            return transformingReader;
+        }
 
-    @Override
-    public DataReader apply(DataReader reader) throws Throwable {
-        TransformingReader transformingReader = new TransformingReader(reader)
-                .add(new UpcaseField(fieldNameList));
-        return transformingReader;
-    }
+        public UpcaseFieldsAction addField(String fieldName) {
+            fieldNameList.add(fieldName);
+            return this;
+        }
 
-    public UpcaseFieldsAction addField(String fieldName) {
-        fieldNameList.add(fieldName);
-        return this;
-    }
+        @Override
+        public void fromRecord(Record source) {
```

Comment:
Let's move the `toRecord()` method above this one to help readability.

Reply:
Done.